devito/propagator.py

```diff
@@ -293,56 +297,74 @@ def generate_space_loops(self, loop_body):
-        :returns: :class:`cgen.Block` representing the loop
+        :returns: :list<cgen.For> a list of for loops
         """
-        ivdep = True
-        remainder = False
+
+        inner_most_dim = True
         orig_loop_body = loop_body
+
         for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
             dim_var = str(self._var_map[spc_var])
             block_var = dim_var + "b"
             loop_limits = self._space_loop_limits[spc_var]
-            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var),
-                                                        block_var),
-                                 dim_var + "<" + block_var+"+"+str(block_size), dim_var + "++", loop_body)
-            if ivdep and len(self.space_dims) > 1:
+
+            if block_size is not None:
+                lower_limit_str = block_var
+                upper_limit_str = block_var + "+" + str(block_size)
+            else:
+                lower_limit_str = str(loop_limits[0])
+                upper_limit_str = str(loop_limits[1])
+
+            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var), lower_limit_str),
+                                 dim_var + "<" + upper_limit_str, dim_var + "++", loop_body)
+
+            if inner_most_dim and len(self.space_dims) > 1:
                 loop_body = cgen.Block([self.compiler.pragma_ivdep] + [loop_body])
-            ivdep = False
+            inner_most_dim = False
 
+        remainder_list = []  # List indicating how many remainder loops we need.
         for spc_var, block_size in reversed(zip(list(self.space_dims), self.block_sizes)):
-            orig_var = str(self._var_map[spc_var])
-            dim_var = orig_var + "b"
-            loop_limits = self._space_loop_limits[spc_var]
-            old_upper_limit = loop_limits[1]
-            new_upper_limit = old_upper_limit-old_upper_limit % block_size
-            if old_upper_limit - new_upper_limit > 0:
-                remainder = True
-            loop_limits = (loop_limits[0], new_upper_limit)
-            loop_body = cgen.For(cgen.InlineInitializer(cgen.Value("int", dim_var),
-                                                        str(loop_limits[0])),
-                                 str(dim_var) + "<" + str(loop_limits[1]), str(dim_var) + "+=" + str(block_size), loop_body)
-        if remainder:
+            # if block size set to None do not block this dimension
+            if block_size is not None:
+                orig_var = str(self._var_map[spc_var])
+                dim_var = orig_var + "b"
+                loop_limits = self._space_loop_limits[spc_var]
+                old_upper_limit = loop_limits[1]                  # sets new upper limit
+                loop_limits = (loop_limits[0], loop_limits[1] - (loop_limits[1] - loop_limits[0]) % block_size)
+
+                if old_upper_limit - loop_limits[1] > 0:  # check old vs new upper limit
```

Comment:
Could you move this condition to immediately following the for loop in line 351 below? That way you don't need to maintain a `remainder_list`. You don't need the additional `for` you introduced in line 347 either this way. You just check each dimension individually on whether it needs a remainder or not (based on this same condition). If it does, you generate the remainder. I think that would also remove both the `if` statements from lines 355 and 361 below.